sdl_viewer/src/main.rs

```diff
@@ -0,0 +1,604 @@
+    moving_right: bool,
+    moving_down: bool,
+    moving_up: bool,
+
+    projection_matrix: Matrix4<f32>,
+    width: i32,
+    height: i32,
+}
+
+impl Camera {
+    fn new(width: i32, height: i32) -> Self {
+        let mut camera = Camera {
+            movement_speed: 1.5,
+            moving_backward: false,
+            moving_forward: false,
+            moving_left: false,
+            moving_right: false,
+            moving_down: false,
+            moving_up: false,
+            mouse_state: MouseState::None,
+            rotate_start: Vector2::zero(),
+            theta: Rad(0.),
+            phi: Rad(0.),
+            transform: Decomposed {
+                scale: 1.,
+                rot: Quaternion::one(),
+                disp: Vector3::new(0., 0., 150.),
+            },
+
+            // These will be set by set_size().
+            projection_matrix: One::one(),
+            width: 0,
+            height: 0,
+        };
+        camera.set_size(width, height);
+        camera
+    }
+
+    fn set_size(&mut self, width: i32, height: i32) {
+        self.width = width;
+        self.height = height;
+        self.projection_matrix =
+            make_projection_matrix(0.1, 10000., Deg(45.), 1., width as f32 / height as f32);
+        unsafe {
+            gl::Viewport(0, 0, width, height);
+        }
+    }
+
+    fn get_world_to_gl(&self) -> Matrix4<f32> {
+        let world_to_camera: Matrix4<f32> = self.transform.inverse_transform().unwrap().into();
+        self.projection_matrix * world_to_camera
+    }
+
+    fn update(&mut self) {
+        let mut pan = Vector3::zero();
+        if self.moving_right {
+            pan.x += 1.;
+        }
+        if self.moving_left {
+            pan.x -= 1.;
+        }
+        if self.moving_backward {
+            pan.z += 1.;
+        }
+        if self.moving_forward {
+            pan.z -= 1.;
+        }
+        if self.moving_up {
+            pan.y += 1.;
+        }
+        if self.moving_down {
+            pan.y -= 1.;
+        }
+
+        if pan.magnitude2() > 0. {
+            let translation = self.transform
+                .rot
+                .rotate_vector(pan.normalize() * self.movement_speed);
+            self.transform.disp += translation;
+        }
+
+        let rotation_z = Quaternion::from_angle_z(self.theta);
+        let rotation_x = Quaternion::from_angle_x(self.phi);
+        self.transform.rot = rotation_z * rotation_x;
+    }
+
+    fn start_rotating(&mut self, x: i32, y: i32) {
+        self.mouse_state = MouseState::Rotating;
+        self.rotate_start.x = x as f32;
+        self.rotate_start.y = y as f32;
+    }
+
+    fn stop_rotating(&mut self) {
+        self.mouse_state = MouseState::None;
+    }
+
+    fn mouse_motion(&mut self, x: i32, y: i32) {
+        if self.mouse_state == MouseState::None {
+            return;
+        }
+        let end = Vector2::new(x as f32, y as f32);
+        let delta = end - self.rotate_start;
```

Comment:
Rename `rotate_start` to `last_mouse_position` and always update? Makes it slightly easier to follow.